### DISCUSSION TITLE
Cannot open osu! through AppImage file

### DISCUSSION THREAD

**Discussion by sevamls** [ppy/osu#27803] [Q&A]
### Type

Crash to desktop

### Bug description

Missing libpthread.so.0. I couldn't find where the directory it reported missing.
I have Fedora Asahi Spin installed (if you need it)

### Version

2024.312.1

### Logs

Dynarec for ARM64, with extension: ASIMD AES CRC32 PMULL SHA1 SHA2 USCAT FLAGM FLAGM2 FRINT PageSize:16384 Running on Icestorm-M1 Firestorm-M1 with 8 Cores
Params database has 38 entries
Box64 with Dynarec v0.2.6 nogit built on Jan 24 2024 00:00:00
BOX64: Detected 48bits at least of address space
Counted 52 Env var
BOX64 LIB PATH: ./:lib/:lib64/:x86_64/:bin64/:libs64/
BOX64 BIN PATH: ./:bin/:/home/meev/.local/bin/:/home/meev/bin/:/usr/local/bin/:/usr/local/sbin/:/usr/bin/:/usr/sbin/:/home/meev/.dotnet/tools/
Looking for /home/meev/Downloads/osu!.AppImage
Rename process to "osu!.AppImage"
Using native(wrapped) libdl.so.2
Using native(wrapped) libpthread.so.0
Using native(wrapped) libz.so.1
Using native(wrapped) libc.so.6
Using native(wrapped) ld-linux-x86-64.so.2
Using native(wrapped) libutil.so.1
Using native(wrapped) librt.so.1
Using native(wrapped) libbsd.so.0
Using native(wrapped) libfuse.so.2
cp: cannot create regular file '/home/meev/.cache/thumbnails/normal/16e3d187be7f97d8a8db7d796233feb2.png': No such file or directory
touch: cannot touch '/home/meev/.local/share/icons/hicolor/.xdg-icon-resource-dummy': No such file or directory
/tmp/.mount_osu!.Apsinqn/usr/bin/osu!: error while loading shared libraries: libpthread.so.0: cannot open shared object file: No such file or directory

UPDATE: I tried to manually copy libpthread into the directory but the "...Apsingn" string changed every time.
UPDATE2: Stuck at libstdc++.so.6/6.0.32. Replacing it will cause the shell to crash.

**Comment by 64ArthurAraujo** [comment]
This is probably better to be converted to a discussion, also i think the .AppImage is for x86_64 only, idk if manually compiling osu to linux-arm would work under m1.

**Comment by sevamls** [comment]
> This is probably better to be converted to a discussion, also i think the .AppImage is for x86_64 only, idk if manually compiling osu to linux-arm would work under m1.

I mean the src build worked (with libraries manually installed). Tbh I don't really mind manually finding libraries and putting them in place.

**Comment by sineplusx** [comment]
> UPDATE: I tried to manually copy libpthread into the directory but the "...Apsingn" string changed every time.

You can extract the game files for a permanent path: `./osu.AppImage --appimage-extract`
The binary for running the game will be extracted to `./squashfs-root/usr/bin/osu!`

Note that the official AppImage release is apparently x64 only.  Would be interested to know if you can actually get the official release to work or if you run into other issues.

**Reply by sevamls** [reply]
`/home/meev/Downloads/squashfs-root/usr/bin/osu!: error while loading shared libraries: libpthread.so.0: cannot open shared object file: No such file or directory
`
I cannot find where my shared libpthread lives.

**Reply by sevamls** [reply]
At this point I'm just going to find and track down each missing library.

**Reply by 64ArthurAraujo** [reply]
I remembered you can use ldd to see where the program is looking for shared libraries
```sh
arthur@fedora:~$ ldd /bin/bash
        linux-vdso.so.1 (0x00007ffce99fd000)
        libtinfo.so.6 => /lib64/libtinfo.so.6 (0x00007f3cdf58d000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f3cdf3ab000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f3cdf743000)
```

**Reply by sevamls** [reply]
> I remembered you can use ldd to see where the program is looking for shared libraries

I used ldd to locate the lib64 folder, only for lib64 to be missing. I corrected the path and everything went smoothly until `libstdc++.so.6` had to be replaced with (I believe) its 64-bit counterpart. Replacing it will cause the shell to crash. Unfortunately, I went a little too close and replaced `libstdc++.so.6.0.32`, bricking the system.
